File: src/lib/ens/cache.ts
// ENS Cache - Dual-layer caching for ENS name resolution
//
// Architecture:
// L1: Memory cache (session-scoped, 1-hour TTL, <1ms latency)
// L2: SQLite cache (persistent, 7-day TTL, ~2-5ms latency)
//
// This reduces RPC calls by >90% and improves user experience

import { invoke } from "@tauri-apps/api/core";

// L1 Memory cache entry
interface MemoryCacheEntry {
	ensName: string | null;
	expiresAt: number;
}

// L2 SQLite cache entry (from Rust)
interface SQLiteCacheEntry {
	address: string;
	ens_name: string | null;
	resolved_at: number;
	expires_at: number;
}

// L1: In-memory cache (session-scoped)
const memoryCache = new Map<string, MemoryCacheEntry>();

// Cache configuration
const MEMORY_TTL_MS = 60 * 60 * 1000; // 1 hour
const SQLITE_TTL_DAYS = 7; // 7 days

/**
 * Get cached ENS name for an address (checks both L1 and L2)
 *
 * @param address - Ethereum address (case-insensitive)
 * @returns ENS name if cached and valid, null if no ENS name, undefined if not cached
 */
export async function getCachedENS(
	address: string,
): Promise<string | null | undefined> {
	const addressLower = address.toLowerCase();

	// 1. Check L1 (Memory) cache - <1ms
	const memoryCached = memoryCache.get(addressLower);
	if (memoryCached && Date.now() < memoryCached.expiresAt) {
		console.log(`🎯 L1 hit for ${address}: ${memoryCached.ensName || "no ENS"}`);
		return memoryCached.ensName;
	}

	// 2. Check L2 (SQLite) cache - ~2-5ms
	try {
		const sqliteCached = await invoke<SQLiteCacheEntry | null>("get_ens_cache", {
			address: addressLower,
		});

		if (sqliteCached) {
			console.log(
				`💾 L2 hit for ${address}: ${sqliteCached.ens_name || "no ENS"}`,
			);

			// Promote to L1 for faster subsequent access
			memoryCache.set(addressLower, {
				ensName: sqliteCached.ens_name,
				expiresAt: Date.now() + MEMORY_TTL_MS,
			});

			return sqliteCached.ens_name;
		}
	} catch (error) {
		console.error("Failed to query L2 cache:", error);
		// Continue to undefined (cache miss)
	}

	// Cache miss
	console.log(`❌ Cache miss for ${address}`);
	return undefined;
}

/**
 * Save ENS name to cache (both L1 and L2)
 *
 * @param address - Ethereum address
 * @param ensName - ENS name (null if address has no ENS)
 */
export async function saveENSCache(
	address: string,
	ensName: string | null,
): Promise<void> {
	const addressLower = address.toLowerCase();

	// Save to L1 (immediate)
	memoryCache.set(addressLower, {
		ensName,
		expiresAt: Date.now() + MEMORY_TTL_MS,
	});

	// Save to L2 (persistent)
	try {
		await invoke("save_ens_cache", {
			address: addressLower,
			ensName,
			ttlDays: SQLITE_TTL_DAYS,
		});
		console.log(`💾 Saved to cache: ${address} -> ${ensName || "no ENS"}`);
	} catch (error) {
		console.error("Failed to save to L2 cache:", error);
	}
}

/**
 * Clear all ENS cache entries (both L1 and L2)
 *
 * @returns Number of L2 entries deleted
 */
export async function clearENSCache(): Promise<number> {
	// Clear L1
	memoryCache.clear();
	console.log("🗑️ Cleared L1 cache");

	// Clear L2
	try {
		const deleted = await invoke<number>("clear_ens_cache");
		console.log(`🗑️ Cleared L2 cache: ${deleted} entries deleted`);
		return deleted;
	} catch (error) {
		console.error("Failed to clear L2 cache:", error);
		return 0;
	}
}

/**
 * Clean up expired ENS cache entries (L2 only, L1 auto-expires)
 *
 * @returns Number of L2 entries deleted
 */
export async function cleanupENSCache(): Promise<number> {
	try {
		const deleted = await invoke<number>("cleanup_ens_cache");
		console.log(`🧹 Cleaned up ${deleted} expired L2 entries`);
		return deleted;
	} catch (error) {
		console.error("Failed to cleanup L2 cache:", error);
		return 0;
	}
}

/**
 * Get ENS cache statistics
 */
export interface ENSCacheStats {
	total_entries: number;
	entries_with_name: number;
	entries_without_name: number;
	expired_entries: number;
	memory_entries: number; // L1 only
}

export async function getENSCacheStats(): Promise<ENSCacheStats> {
	try {
		const sqliteStats = await invoke<Omit<ENSCacheStats, "memory_entries">>(
			"get_ens_cache_stats",
		);

		return {
			...sqliteStats,
			memory_entries: memoryCache.size,
		};
	} catch (error) {
		console.error("Failed to get cache stats:", error);
		return {
			total_entries: 0,
			entries_with_name: 0,
			entries_without_name: 0,
			expired_entries: 0,
			memory_entries: memoryCache.size,
		};
	}
}

File: src-tauri/src/ens/cache.rs
//! ENS name cache module
//!
//! This module provides SQLite-backed caching for ENS reverse resolution results.
//! Caching reduces RPC calls and improves performance when displaying ENS names.

use crate::db;
use serde::{Deserialize, Serialize};

/// ENS cache entry
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct ENSCache {
    pub address: String,
    pub ens_name: Option<String>,
    pub resolved_at: i64,
    pub expires_at: i64,
}

/// Get cached ENS name for an address
///
/// Returns None if:
/// - No cache entry exists
/// - Cache entry has expired
pub async fn get_cached_ens(address: &str) -> Result<Option<ENSCache>, String> {
    let pool = db::pool();
    let now = chrono::Utc::now().timestamp();

    // Normalize address to lowercase for consistent lookups
    let address_lower = address.to_lowercase();

    let result = sqlx::query_as::<_, ENSCache>(
        "SELECT * FROM ens_cache
         WHERE address = ? AND expires_at > ?",
    )
    .bind(&address_lower)
    .bind(now)
    .fetch_optional(pool.as_ref())
    .await
    .map_err(|e| format!("Failed to query ENS cache: {}", e))?;

    Ok(result)
}

/// Save ENS name to cache
///
/// # Arguments
/// * `address` - Ethereum address (will be normalized to lowercase)
/// * `ens_name` - ENS name (None if address has no ENS)
/// * `ttl_days` - Time to live in days (default: 7)
pub async fn save_ens_cache(
    address: &str,
    ens_name: Option<&str>,
    ttl_days: i64,
) -> Result<(), String> {
    let pool = db::pool();
    let now = chrono::Utc::now().timestamp();
    let expires_at = now + (ttl_days * 24 * 60 * 60);

    // Normalize address to lowercase
    let address_lower = address.to_lowercase();

    sqlx::query(
        "INSERT INTO ens_cache (address, ens_name, resolved_at, expires_at)
         VALUES (?, ?, ?, ?)
         ON CONFLICT(address) DO UPDATE SET
           ens_name = excluded.ens_name,
           resolved_at = excluded.resolved_at,
           expires_at = excluded.expires_at",
    )
    .bind(&address_lower)
    .bind(ens_name)
    .bind(now)
    .bind(expires_at)
    .execute(pool.as_ref())
    .await
    .map_err(|e| format!("Failed to save ENS cache: {}", e))?;

    Ok(())
}

/// Clean up expired ENS cache entries
///
/// Returns the number of deleted entries
pub async fn cleanup_expired_ens_cache() -> Result<u64, String> {
    let pool = db::pool();
    let now = chrono::Utc::now().timestamp();

    let result = sqlx::query("DELETE FROM ens_cache WHERE expires_at <= ?")
        .bind(now)
        .execute(pool.as_ref())
        .await
        .map_err(|e| format!("Failed to cleanup ENS cache: {}", e))?;

    Ok(result.rows_affected())
}

/// Clear all ENS cache entries
///
/// Returns the number of deleted entries
pub async fn clear_all_ens_cache() -> Result<u64, String> {
    let pool = db::pool();

    let result = sqlx::query("DELETE FROM ens_cache")
        .execute(pool.as_ref())
        .await
        .map_err(|e| format!("Failed to clear ENS cache: {}", e))?;

    Ok(result.rows_affected())
}

/// Get ENS cache statistics
#[derive(Debug, Serialize, Deserialize)]
pub struct ENSCacheStats {
    pub total_entries: i64,
    pub entries_with_name: i64,
    pub entries_without_name: i64,
    pub expired_entries: i64,
}

pub async fn get_ens_cache_stats() -> Result<ENSCacheStats, String> {
    let pool = db::pool();
    let now = chrono::Utc::now().timestamp();

    let total_entries = sqlx::query_as::<_, (i64,)>("SELECT COUNT(*) FROM ens_cache")
        .fetch_one(pool.as_ref())
        .await
        .map_err(|e| format!("Failed to count total entries: {}", e))?
        .0;

    let entries_with_name =
        sqlx::query_as::<_, (i64,)>("SELECT COUNT(*) FROM ens_cache WHERE ens_name IS NOT NULL")
            .fetch_one(pool.as_ref())
            .await
            .map_err(|e| format!("Failed to count entries with name: {}", e))?
            .0;

    let entries_without_name =
        sqlx::query_as::<_, (i64,)>("SELECT COUNT(*) FROM ens_cache WHERE ens_name IS NULL")
            .fetch_one(pool.as_ref())
            .await
            .map_err(|e| format!("Failed to count entries without name: {}", e))?
            .0;

    let expired_entries =
        sqlx::query_as::<_, (i64,)>("SELECT COUNT(*) FROM ens_cache WHERE expires_at <= ?")
            .bind(now)
            .fetch_one(pool.as_ref())
            .await
            .map_err(|e| format!("Failed to count expired entries: {}", e))?
            .0;

    Ok(ENSCacheStats {
        total_entries,
        entries_with_name,
        entries_without_name,
        expired_entries,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_ens_cache() {
        // This test requires database initialization
        // Run with: cargo test --package colimail --lib ens::cache::tests::test_ens_cache
    }
}
